cmd/neofs-node/meta.go

```diff
@@ -0,0 +1,133 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+
+	"github.com/nspcc-dev/neofs-node/pkg/core/container"
+	"github.com/nspcc-dev/neofs-node/pkg/core/netmap"
+	cntClient "github.com/nspcc-dev/neofs-node/pkg/morph/client/container"
+	"github.com/nspcc-dev/neofs-node/pkg/services/meta"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	netmapsdk "github.com/nspcc-dev/neofs-sdk-go/netmap"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+func initMeta(c *cfg) {
+	if c.cfgMorph.client == nil {
+		initMorphComponents(c)
+	}
+
+	c.cfgMeta.cLister = &containerListener{
+		key:        c.binPublicKey,
+		cnrClient:  c.basics.cCli,
+		containers: c.cfgObject.cnrSource,
+		network:    c.basics.netMapSource,
+	}
+
+	m, err := meta.New(c.log.With(zap.String("service", "meta data")),
+		c.cfgMeta.cLister,
+		c.applicationConfiguration.fsChain.dialTimeout,
+		c.applicationConfiguration.fsChain.endpoints,
+		c.basics.containerSH,
+		c.basics.netmapSH,
+		c.applicationConfiguration.metadata.path)
+	fatalOnErr(err)
+
+	c.workers = append(c.workers, newWorkerFromFunc(func(ctx context.Context) {
+		err = m.Run(ctx)
+		if err != nil {
+			c.internalErr <- fmt.Errorf("meta data service error: %w", err)
+		}
+	}))
+}
+
+type containerListener struct {
+	key []byte
+
+	cnrClient  *cntClient.Client
+	containers container.Source
+	network    netmap.Source
+
+	m          sync.RWMutex
+	prevCnrs   []cid.ID
+	prevNetMap *netmapsdk.NetMap
+	prevRes    map[cid.ID]struct{}
+}
+
+func (c *containerListener) List() (map[cid.ID]struct{}, error) {
+	actualContainers, err := c.cnrClient.List(nil)
+	if err != nil {
+		return nil, fmt.Errorf("read containers: %w", err)
+	}
+	curEpoch, err := c.network.Epoch()
+	if err != nil {
+		return nil, fmt.Errorf("read current NeoFS epoch: %w", err)
+	}
+	networkMap, err := c.network.GetNetMapByEpoch(curEpoch)
+	if err != nil {
+		return nil, fmt.Errorf("read network map at the current epoch #%d: %w", curEpoch, err)
+	}
+
+	c.m.RLock()
+	if c.prevNetMap != nil {
+		cnrsSame := slices.EqualFunc(c.prevCnrs, actualContainers, func(cID1, cID2 cid.ID) bool {
+			return bytes.Equal(cID1[:], cID2[:])
+		})
+		netmapSame := slices.EqualFunc(c.prevNetMap.Nodes(), networkMap.Nodes(), func(n1 netmapsdk.NodeInfo, n2 netmapsdk.NodeInfo) bool {
```

Comment:
seems like no need to calc this if `!cnrsSame`

Reply:
ok but required one more `m.Unlock` which i usually do not like